Networking/Networking/ApplicationPassword/RequestProcessor.swift

```diff
@@ -0,0 +1,96 @@
+import Alamofire
+import Foundation
+
+/// Authenticates and retries requests
+///
+final class RequestProcessor {
+    private var requestsToRetry = [RequestRetryCompletion]()
+
+    private var isAuthenticating = false
+
+    private let requestAuthenticator: RequestAuthenticator
+
+    init(credentials: Credentials?) {
+        requestAuthenticator = RequestAuthenticator(credentials: credentials)
+    }
+}
+
+// MARK: Request Authentication
+//
+extension RequestProcessor: RequestAdapter {
+    func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        return try requestAuthenticator.authenticate(urlRequest)
+    }
+}
+
+// MARK: Retrying Request
+//
+extension RequestProcessor: RequestRetrier {
+    func should(_ manager: Alamofire.SessionManager,
+                retry request: Alamofire.Request,
+                with error: Error,
+                completion: @escaping Alamofire.RequestRetryCompletion) {
+        guard
+            request.retryCount == 0, // Only retry once
+            shouldRetry(request), // Retry only REST API requests that use application password
+            shouldRetry(error) // Retry only specific errors
+        else {
+            return completion(false, 0.0)
+        }
+
+        requestsToRetry.append(completion)
+        if !isAuthenticating {
+            generateApplicationPassword()
+        }
+    }
+}
+
+// MARK: Helpers
+//
+private extension RequestProcessor {
+    func generateApplicationPassword() {
+        Task(priority: .medium) {
+            isAuthenticating = true
+
+            do {
+                let _ = try await requestAuthenticator.generateApplicationPassword()
+                isAuthenticating = false
+                completeRequests(true)
+            } catch {
+                isAuthenticating = false
+                completeRequests(false)
+            }
+        }
+    }
+
+    func shouldRetry(_ request: Alamofire.Request) -> Bool {
+        guard case let .wporg(_, _, siteAddress) = requestAuthenticator.credentials,
+              let url = request.request?.url,
+              url.absoluteString.hasPrefix(siteAddress + "/" + RESTRequest.Settings.basePath) else {
```

Comment:
these lines seem to be exactly the same as in `RequestAuthenticator.authenticate`, can this check be moved to `RequestAuthenticator` and called here?

Reply:
Good idea. Done in https://github.com/woocommerce/woocommerce-ios/pull/8483/commits/4f932ef0417660e204de65e11d82bb37ebb840f5